File: Src/PRuntimes/PSymRuntime/src/main/java/psym/runtime/values/PSeq.java
package psym.runtime.values;

import java.util.ArrayList;
import java.util.List;
import psym.runtime.values.exceptions.InvalidIndexException;

public class PSeq extends PCollection {
  // stores the map
  private final List<PValue<?>> seq;

  public PSeq(List<PValue<?>> input_seq) {
    seq = new ArrayList<>();
    for (PValue<?> entry : input_seq) {
      seq.add(PValue.clone(entry));
    }
  }

  public PSeq(PSeq other) {
    seq = new ArrayList<>();
    for (PValue<?> entry : other.seq) {
      seq.add(PValue.clone(entry));
    }
  }

  public PValue<?> getValue(int index) throws InvalidIndexException {
    if (index >= seq.size() || index < 0) throw new InvalidIndexException(index, this);
    return seq.get(index);
  }

  public void setValue(int index, PValue<?> val) throws InvalidIndexException {
    if (index >= seq.size() || index < 0) throw new InvalidIndexException(index, this);
    seq.set(index, val);
  }

  public void insertValue(int index, PValue<?> val) throws InvalidIndexException {
    if (index > seq.size() || index < 0) throw new InvalidIndexException(index, this);
    seq.add(index, val);
  }

  @Override
  public PSeq clone() {
    return new PSeq(seq);
  }

  @Override
  public int hashCode() {
    return ComputeHash.getHashCode(seq);
  }

  @Override
  public boolean equals(Object obj) {
    if (obj == this) return true;

    if (!(obj instanceof PSeq)) {
      return false;
    }

    PSeq other = (PSeq) obj;
    if (seq.size() != other.seq.size()) {
      return false;
    }

    for (int i = 0; i < seq.size(); i++) {
      if (!PValue.equals(other.seq.get(i), this.seq.get(i))) {
        return false;
      }
    }
    return true;
  }

  @Override
  public String toString() {
    StringBuilder sb = new StringBuilder();
    sb.append("[");
    String sep = "";
    for (PValue<?> item : seq) {
      sb.append(sep);
      sb.append(item);
      sep = ", ";
    }
    sb.append("]");
    return sb.toString();
  }

  @Override
  public int size() {
    return seq.size();
  }

  @Override
  public boolean contains(PValue<?> item) {
    return seq.contains(item);
  }
}

File: Src/PChecker/CheckerCore/Runtime/Values/PSeq.cs
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using PChecker.Runtime.Values.Exceptions;

namespace PChecker.Runtime.Values
{
    public sealed class PSeq : IPMutableValue, IReadOnlyList<IPValue>
    {
        private readonly List<IPValue> values = new List<IPValue>();

        private int hashCode;
        private bool isDirty;
        private bool isFrozen;

        public PSeq()
        {
            hashCode = ComputeHashCode();
        }

        public PSeq(IEnumerable<IPValue> values)
        {
            this.values = values.ToList();
            hashCode = ComputeHashCode();
        }

        private bool IsDirty
        {
            get => isDirty;
            set
            {
                if (value && isFrozen)
                {
                    throw new PFrozenMutationException();
                }

                isDirty = value;
            }
        }

        public void Freeze()
        {
            foreach (var value in values)
            {
                MutabilityHelper.EnsureFrozen(value);
            }

            isFrozen = true;
        }

        public IPValue Clone()
        {
            return new PSeq(values.Select(item => item?.Clone()));
        }

        public bool Equals(IPValue other)
        {
            return other is PSeq otherValue && Equals(otherValue);
        }

        public IEnumerator<IPValue> GetEnumerator()
        {
            return values.GetEnumerator();
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return GetEnumerator();
        }

        public int Count => values.Count;

        public IPValue this[int index]
        {
            get => values[index];
            set
            {
                if (isFrozen)
                {
                    throw new PFrozenMutationException();
                }

                values[index] = value;
            }
        }

        private int ComputeHashCode()
        {
            return HashHelper.ComputeHash(values);
        }

        public void Add(IPValue item)
        {
            IsDirty = true;
            values.Add(item?.Clone());
        }

        public void Insert(int index, IPValue item)
        {
            IsDirty = true;
            values.Insert(index, item?.Clone());
        }

        public void RemoveAt(int index)
        {
            IsDirty = true;
            values.RemoveAt(index);
        }

        public bool Contains(IPValue item)
        {
            return values.Contains(item);
        }

        private bool Equals(PSeq other)
        {
            return other != null && values.SequenceEqual(other.values);
        }

        public override int GetHashCode()
        {
            if (!IsDirty)
            {
                return hashCode;
            }

            hashCode = ComputeHashCode();
            IsDirty = false;

            return hashCode;
        }

        public override string ToString()
        {
            var sb = new StringBuilder();
            sb.Append("(");
            var sep = "";
            foreach (var value in values)
            {
                sb.Append(sep);
                sb.Append(value);
                sep = ", ";
            }

            sb.Append(")");
            return sb.ToString();
        }

        public string ToEscapedString()
        {
            var sb = new StringBuilder();
            sb.Append("(");
            var sep = "";
            foreach (var value in values)
            {
                var v = value == null ? "null" : value.ToEscapedString();
                sb.Append(sep);
                sb.Append(v);
                sep = ", ";
            }

            sb.Append(")");
            return sb.ToString();
        }

        public object ToDict()
        {
            return values.Select(value => value == null ? null : value.ToDict()).ToList();
        }
    }
}